### PR TITLE
Arcade build scripts

### PR DESCRIPTION
**Pull request by tmat**
Use more of Arcade bootstrapper functions.
Convert buildCoreClr to msbuildEngine parameter used by Arcade.
Update Initialize* functions to memoize their result.

Use build.ps1 only for builds that follow standard build pipeline structure (restore, build, pack, test, sign, publish). Add a separate script for testing determinism and move common functions down to build-utils.ps1.

Use standard build pipeline from Arcade (build.proj).

Change `build.sh` to match `build.ps1` wherever possible.

### CODE REVIEW

**Comment by tmat** [comment]
@jaredpar PTAL

File: .gitignore [hunk 1]
```diff
@@ -17,6 +17,7 @@ x64/
 [Oo]bj/
 .dotnet/
 .tools/
+.packages/
```
Comment (jaredpar) [inline .gitignore]: Why is this necessary now?

File: build/scripts/build-utils.ps1 [hunk 2]
```diff
@@ -6,10 +6,16 @@ $ErrorActionPreference="Stop"
 # Import Arcade functions
 . (Join-Path $PSScriptRoot "tools.ps1")
 
-[string]$binariesDir = Join-Path $RepoRoot "Binaries"
+$BinariesConfigDir = Join-Path $ArtifactsDir $configuration # TODO: remove
```
Comment (jaredpar) [inline build/scripts/build-utils.ps1]: Is there a bug tracking this TODO?

File: build/scripts/build.ps1 [hunk 3]
```diff
@@ -17,73 +17,88 @@
...
+    Write-Host "  -bootstrap                Build using a bootstrap compilers"
+    Write-Host "  -msbuildEngine <value>    Msbuild engine to use to run build ('dotnet', 'vs', or unspecified)."
+    Write-Host "  -procdump                 Monitor test runs with procdump"
+    Write-Host "  -skipAnalyzers            Do not run analyzers during build operations"
+    Write-Host "  -prepareMachine           Prepare machine for CI run, clean up processes after build"
+    Write-Host ""
+    Write-Host "Command line arguments starting with '/p:' are passed through to MSBuild."
+}
+
+if ($help -or (($properties -ne $null) -and ($properties.Contains("/help") -or $properties.Contains("/?")))) {
```
Comment (jaredpar) [inline build/scripts/build.ps1]: Why put this here vs. the main `try / finally` body?

File: .gitignore (hunk 1 above)
Comment (tmat) [reply .gitignore]: On CI Arcade restores packages to this dir to isolate.
Do we want to do that for all legs or just one to save restore time?

File: build/scripts/build-utils.ps1 (hunk 2 above)
Comment (tmat) [reply build/scripts/build-utils.ps1]: Working on that (changing the layout) now.

File: build/scripts/build.ps1 (hunk 3 above)
Comment (tmat) [reply build/scripts/build.ps1]: Will move.

File: build/scripts/build-utils.ps1 [hunk 4]
```diff
@@ -6,10 +6,16 @@ $ErrorActionPreference="Stop"
...
+$VSSetupDir = Join-Path $ArtifactsDir (Join-Path "VSSetup" $configuration)
+
+$binaryLog = if (Test-Path variable:binaryLog) { $binaryLog } else { $false }
+$nodereuse = if (Test-Path variable:nodereuse) { $nodereuse } else { $false }
+$bootstrapDir = if (Test-Path variable:bootstrapDir) { $bootstrapDir } else { "" }
+$properties = if (Test-Path variable:properties) { $properties } else { @() }
```
[the remark this answers is not in the document]
Comment (tmat) [reply build/scripts/build-utils.ps1]: These are script parameters.

File: build/scripts/build.ps1 [hunk 5]
```diff
@@ -17,73 +17,88 @@
...
-    Write-Host "Special Test options"
-    Write-Host "  -testDeterminism          Run determinism tests"
+    Write-Host "Advanced settings:"
+    Write-Host "  -ci                       Set when running on CI server"
+    Write-Host "  -official                 Set when building an official build"
+    Write-Host "  -bootstrap                Build using a bootstrap compilers"
+    Write-Host "  -msbuildEngine <value>    Msbuild engine to use to run build ('dotnet', 'vs', or unspecified)."
+    Write-Host "  -procdump                 Monitor test runs with procdump"
+    Write-Host "  -skipAnalyzers            Do not run analyzers during build operations"
+    Write-Host "  -prepareMachine           Prepare machine for CI run, clean up processes after build"
```
[the remark this answers is not in the document]
Comment (tmat) [reply build/scripts/build.ps1]: In `tools.ps1` when terminating build processes.

File: build/ToolsetPackages/RoslynToolset.csproj [hunk 6]
```diff
@@ -1,42 +0,0 @@
...
-    <PackageReference Include="RoslynTools.ModifyVsixManifest" Version="$(RoslynToolsModifyVsixManifestVersion)" ExcludeAssets="all" />
-    <PackageReference Include="RoslynTools.OptProf" Version="$(RoslynToolsOptProfVersion)" ExcludeAssets="all" />
-    <PackageReference Include="Roslyn.OptProf.RunSettings.Generator" Version="$(RoslynOptProfRunSettingsGeneratorVersion)" ExcludeAssets="all" />
-    <PackageReference Include="RoslynTools.MSBuild" Version="$(RoslynToolsMSBuildVersion)" ExcludeAssets="all" />
-    <PackageReference Include="xunit.runner.wpf" Version="$(xunitrunnerwpfVersion)" ExcludeAssets="all" />
-    <PackageReference Include="vswhere" Version="$(vswhereVersion)" ExcludeAssets="all" />
-  </ItemGroup>
-  <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp2.1'">
-    <PackageReference Include="Microsoft.NETCore.Compilers" Version="$(MicrosoftNETCoreCompilersVersion)" ExcludeAssets="all" />
-  </ItemGroup>
```
[the remark this answers is not in the document]
Comment (tmat) [reply build/ToolsetPackages/RoslynToolset.csproj]: Tools are restored by Arcade (currently in `RepoToolset\tools.proj`). 
Repo can add additional packages in `Tools.props`.

File: build/scripts/build-utils.ps1 (hunk 4 above)
[the remark this answers is not in the document]
Comment (tmat) [reply build/scripts/build-utils.ps1]: I'll fix `nodeReuse` capitalization.

File: build/scripts/build-utils.ps1 (hunk 2 above)
Comment (tmat) [reply build/scripts/build-utils.ps1]: https://github.com/dotnet/roslyn/pull/31560

File: build/scripts/build.ps1 (hunk 3 above)
Comment (tmat) [reply build/scripts/build.ps1]: Moved to Process-Arguments

File: .gitignore (hunk 1 above)
Comment (tmat) [reply .gitignore]: Switched to using global nuget cache, except for in test-correctness.ps1, where it restores to `.packages` dir.
